base/tps/src/org/dogtagpki/server/tps/rest/ActivityService.java

```diff
@@ -136,24 +159,70 @@ protected void retrieveActivitiesWithVLV(
             Integer size,
             ActivityCollection response) throws Exception {
 
+        String method = "ActivityService.retrieveActivitiesWithVLV: ";
+        CMS.debug(method);
         // search with VLV sorted by date in reverse order
         IDBVirtualList<ActivityRecord> list = database.findRecords(
                 null, null, new String[] { "-date" }, size);
 
+        List<String> authorizedProfiles = getAuthorizedProfiles();
+
         int total = list.getSize();
+        CMS.debug(method + "total: " + total);
+        int retTotal = 0; // debugging only
 
         // return entries in the requested page
-        for (int i = start; i < start + size && i < total; i++) {
-            ActivityRecord record = list.getElementAt(i);
-
-            if (record == null) {
-                CMS.debug("ActivityService: Activity record not found");
-                throw new PKIException("Activity record not found");
+        if (authorizedProfiles != null) {
+            if (authorizedProfiles.contains(UserResource.ALL_PROFILES)) {
+                for (int i = start; i < start + size && i < total; i++) {
+                    ActivityRecord record = list.getElementAt(i);
+
+                    response.addEntry(createActivityData(record));
+                }
+                retTotal = total;
+            } else { // not authorized for all profiles
+                for (int i = start; i < start + size && i < total; i++) {
+                    ActivityRecord record = list.getElementAt(i);
+
+                    //CMS.debug(method + "record.Id="+ record.getId());
+                    // for some reason, not all activities have token type
+                    // filled in. It is therefore necessary to get it from
+                    // the token record directly.
+                    String type = record.getType();
+                    //CMS.debug(method + "record.tokenType="+ type);
+                    if ((type == null) || type.isEmpty()) {
+                        CMS.debug(method + "record.tokenType null...getting from token record");
+                        String tokenID = record.getTokenID();
+                        if ((tokenID == null) || (tokenID.isEmpty())) {
+                            CMS.debug(method + "record tokenID not found");
+                            continue;
+                        }
+                        TPSSubsystem subsystem = (TPSSubsystem) CMS.getSubsystem(TPSSubsystem.ID);
+                        TokenDatabase t_database = subsystem.getTokenDatabase();
+                        TokenRecord t_record = t_database.getRecord(tokenID);
+                         if (t_record == null) {
+                            CMS.debug(method + "token record not found");
+                            continue;
+                            }
+                        type = t_record.getType();
+                    }
```

Comment:
The `continue` statements mean that the activity record will not be included in the `response`, but it will still counted in the `total`, and might affect page size too. Should it be like this instead?
```
if (type == null || !type.isEmpty()) {
    String tokenID = record.getTokenID();
    if (tokenID != null && !tokenID.isEmpty()) {
        TokenRecord t_record = t_database.getRecord(tokenID);
        if (t_record != null) {
            type = t_record.getType();
        }
    }
}
```

Reply:
Good catch!  Yes we should just keep type null and let it be added to the response. Thanks!